src/Stratis.Bitcoin/P2P/PeerDiscoveryLoop.cs

```diff
@@ -0,0 +1,133 @@
+            Guard.NotNull(asyncLoopFactory, nameof(asyncLoopFactory));
+
+            this.applicationStopping = applicationStopping;
+            this.asyncLoopFactory = asyncLoopFactory;
+            this.parameters = parameters;
+            this.addressManager = parameters.PeerAddressManager();
+            this.network = network;
+            this.peersToFind = this.parameters.PeerAddressManagerBehaviour().PeersToDiscover;
+        }
+
+        public void Start()
+        {
+            this.asyncLoop = this.asyncLoopFactory.Run(nameof(this.DiscoverPeersAsync), async token =>
+            {
+                await DiscoverPeersAsync().ConfigureAwait(false);
+            },
+            this.applicationStopping,
+            TimeSpans.RunOnce);
+        }
+
+        private async Task DiscoverPeersAsync()
+        {
+            while (this.addressManager.Peers.Count < this.peersToFind)
+            {
+                this.applicationStopping.ThrowIfCancellationRequested();
+
+                var peersToDiscover = new List<NetworkAddress>();
+                peersToDiscover.AddRange(this.addressManager.SelectPeersToConnectTo());
+
+                if (peersToDiscover.Count == 0)
+                {
+                    PopulateTableWithDNSNodes(peersToDiscover);
+                    PopulateTableWithHardNodes(peersToDiscover);
+
+                    peersToDiscover = new List<NetworkAddress>(peersToDiscover.OrderBy(a => RandomUtils.GetInt32()));
+                    if (peersToDiscover.Count == 0)
+                        await Task.CompletedTask;
+                }
+
+                var clonedParameters = this.parameters.Clone();
+
+                Parallel.ForEach(peersToDiscover, new ParallelOptions()
+                {
+                    MaxDegreeOfParallelism = 5,
+                    CancellationToken = this.applicationStopping,
+                },
+                peer =>
+                {
+                    using (var connectTokenSource = CancellationTokenSource.CreateLinkedTokenSource(this.applicationStopping))
+                    {
+                        connectTokenSource.CancelAfter(TimeSpan.FromSeconds(5));
+
+                        Node node = null;
+
+                        try
+                        {
+                            clonedParameters.ConnectCancellation = connectTokenSource.Token;
+
+                            var addressManagerBehaviour = clonedParameters.TemplateBehaviors.Find<PeerAddressManagerBehaviour>();
+                            clonedParameters.TemplateBehaviors.Clear();
+                            clonedParameters.TemplateBehaviors.Add(addressManagerBehaviour);
+
+                            node = Node.Connect(this.network, peer.Endpoint, clonedParameters);
+                            node.VersionHandshake(connectTokenSource.Token);
+                            node.SendMessageAsync(new GetAddrPayload());
+
+                            connectTokenSource.Token.WaitHandle.WaitOne(2000);
```

Comment:
Is `2000` this really enough?
What if the connection is a bit slow then we did not have time to receive the addresses.

Reply:
OK yeah, I have increased this to 5000.

Reply:
Also @dangershony why do we need to do:

```node.SendMessageAsync(new GetAddrPayload());```

after we have successfully handshaked? Remember we are only doing discovery here?